### ISSUE TITLE
Usage Reporting for Syncthing Wrappers

### ISSUE THREAD

**Issue by Nutomic** [syncthing/syncthing#3672]
I've been thinking about adding for wrappers (eg the Android app). We could put these stats into a seperate category for each wrapper. Some things that would be interesting:
- CPU architecture
- always run in background / only run on wifi / only run on battery / restrict to wifi networks
- using root
- use tor
- keep CPU awake
- run with foreground priority

What do you think about this? The implementation could be done with an API like `POST /usage-report`, that is automatically called by the wrapper every day or so. Related to #3628

**Comment by canton7** [comment]
Those stats are very android-specific...

A more generic 'wrapper stats' might make more sense, but it would be hard to parse for data.syncthing.net, and it would be hard to get a user's consent for it...

**Comment by Nutomic** [comment]
Right, my idea was to have seperate stats for every wrapper. Having the same ones doesn't really make sense.

**Comment by calmh** [comment]
I could see a call to for example "register" and "external" feature or some such, as a set of string. You would do something like `POST /rest/system/external-tags` with a JSON array `["android", "runningAsRoot", "usingTor"]` or whatever you like. This could be passed on verbatim to the usage reporting server and result in something like the current feature usage table, per keyword. So you'd see how many use inotify, synctrayzor, rooted phones, whatevs someone wants to inject into the system.

**Comment by Nutomic** [comment]
It should probably be a key:value map, but otherwise it sounds good.

Also should probably update the urAccepted value whenever these "external feature" keys change.

**Comment by calmh** [comment]
That sounds complicated and annoying so I'm not convinced, both the arbitrarily typed map and reacting to external changes like that. We don't revoke urAccepted because values change, normally.

**Comment by canton7** [comment]
An array sounds like a hard way to gather statistics with numbers in though - `{"numWatchedFolders": 3}`, is easier to parse than `["numWatchedFolders-3"]`

**Comment by calmh** [comment]
Indeed. But allowing arbitrary value types (rather than just booleans as I first imagined) makes it a rather larger thing to support, not least server-side.

**Comment by AudriusButkevicius** [comment]
Why? I guess we could ignore values that we don't have columns for, and the wrapper devs would have to ask columns added ahead of time.

**Comment by calmh** [comment]
Sort of defeats the point. I don't want to be a manual link in that chain, at all. :D

**Comment by calmh** [comment]
iPad misclick

**Comment by xor-gate** [comment]
It would be nice to inject a "user-agent" style string into the stats, just one string, nothing more. As I'm the maintainer of [`syncthing-macosx`](https://github.com/xor-gate/syncthing-macosx) package and it would be nice to see how many users it actually use. Something like the format in [RFC2616 - 14.43 User Agent](https://tools.ietf.org/html/rfc2616#section-14.43)

**Comment by AudriusButkevicius** [comment]
Alternatively, we could temporarily store column we are not aware of values somewhere on disk for a day, and if we get enough (let's say 30) submissions with that data point, we automatically infer the time, add the column, and backfill the column in SQL.

**Comment by calmh** [comment]
Gnnh. Sure. Or move to an unstructured storage backend, say riak or mongodb or something. It's all just programming, of course, but in practice it'll be the difference between "I can do that tonight" and "yeah maybe next year or not at all" :) (For me at least, not speaking for anyone else)

**Comment by Nutomic** [comment]
@calmh I misunderstood your reply, disregard what I said about urAccepted.

As for the data, boolean values should be fine (at least for now). Still, I would prefer to have the json formatted as key:value, because it's easier to parse. I certainly don't want to make this more complicated than it needs to be :)

**Comment by calmh** [comment]
Yeah no I agree, that would be nice to be able to add arbitrary fields with values. But as a quick interim solution, all the things in the original request up top could be modeled as a set of strings denoting which extra/external features are in use. Sure, "cpu architecture" is an odd "feature" but we could still handle it that way. @canton7's example of "number of watched folders" doesn't fit into that though.

**Comment by Nutomic** [comment]
Like I said, boolean values would be totally fine. I'm just disagreeing how the JSON should be formatted :p 

Things like cpu architecture or number of folders are included in the normal usage report anyway. So those aren't really important (and could be done on the server if we send the wrapper name).

**Comment by canton7** [comment]
What database do you use? Some, like postgres, support columns of unstructured data.
